### PR TITLE
Use simple diffs except in `DiffSeq`

### LINKED ISSUE #1114: Use `Simple` diffs in V2 InMemory
An ad-hoc integration resulting in this:
```
Just adding
  10000 transactions
    setup mempool:             OK (0.16s)
      36.7 μs ± 3.4 μs, 1.6 KB allocated,   1 B  copied,  10 MB peak memory
    setup mempool + benchmark: OK (0.10s)
      33.1 ms ± 3.0 ms, 107 MB allocated, 4.0 MB copied,  12 MB peak memory
    test:                      OK (0.03s)
    txs length:                OK
  20000 transactions
    setup mempool:             OK (0.39s)
      84.6 μs ± 6.0 μs, 1.6 KB allocated,   1 B  copied,  18 MB peak memory
    setup mempool + benchmark: OK (0.99s)
      64.7 ms ± 4.7 ms, 227 MB allocated, 8.6 MB copied,  27 MB peak memory
    test:                      OK (0.07s)
    txs length:                OK

All 8 tests passed (1.75s)
```

compared with the results in `utxo-hd-8.9`
```
❯ cabal run mempool-bench -- +RTS -T
Resolving dependencies...
Just adding
  10000 transactions
    setup mempool:             OK (0.30s)
      34.7 μs ± 1.8 μs, 2.0 KB allocated,   1 B  copied,  10 MB peak memory
    setup mempool + benchmark: OK (2.27s)
      756  ms ±  18 ms, 2.0 GB allocated, 6.8 MB copied,  14 MB peak memory
    test:                      OK (0.77s)
    txs length:                OK
  20000 transactions
    setup mempool:             OK (0.20s)
      82.4 μs ± 6.0 μs,   0 B  allocated,   0 B  copied,  18 MB peak memory
    setup mempool + benchmark: OK (8.87s)
      2.952 s ±  60 ms, 7.7 GB allocated,  16 MB copied,  23 MB peak memory
    test:                      OK (2.93s)
    txs length:                OK

All 8 tests passed (15.34s)
```

We have to incorporate these changes into the UTXO-HD branch https://github.com/input-output-hk/anti-diffs/pull/49

### PR DESCRIPTION
**Pull request by jasagredo**
Closes #1114

### CODE REVIEW

File: ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs [hunk 1]
```diff
@@ -271,10 +270,10 @@ shelleyAvvmAddressesAreDeletesInUtxoDiff srcLedgerState destLedgerState =
   where
     toNextUtxoDiff
       :: LedgerState (ShelleyBlock Proto (ShelleyEra Crypto)) EmptyMK
-      -> Diff (TxIn Crypto) (Core.TxOut (AllegraEra Crypto))
+      -> Diff.Diff (TxIn Crypto) (Core.TxOut (AllegraEra Crypto))
     toNextUtxoDiff = avvmAddressesToUtxoDiff . stashedAVVMAddresses . shelleyLedgerState
     avvmAddressesToUtxoDiff (UTxO m) =
-      let func txOut = Diff.singletonDelete (Core.translateEra' () txOut)
+      let func _txOut = Diff.Delete -- (Core.translateEra' () txOut)
```
Comment (jasagredo) [inline ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs:276]: Remove this comment

File: ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs [hunk 2]
```diff
@@ -304,7 +303,7 @@ nonEmptyAvvmAddresses ledgerState =
 
 extractUtxoDiff
   :: LedgerState (ShelleyBlock proto era) DiffMK
-  -> Diff (TxIn (EraCrypto era)) (Core.TxOut era)
+  -> Diff.Diff (TxIn (EraCrypto era)) (Core.TxOut era)
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs:306]:
```suggestion
  -> Diff (TxIn (EraCrypto era)) (Core.TxOut era)
```

File: ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs [hunk 3]
```diff
@@ -53,6 +51,7 @@ import           Ouroboros.Consensus.HardFork.Combinator.State.Types
 import           Ouroboros.Consensus.Ledger.Basics (LedgerCfg, LedgerConfig,
                      LedgerState)
 import           Ouroboros.Consensus.Ledger.Tables
+import qualified Ouroboros.Consensus.Ledger.Tables.Diffs as Diff
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/test/cardano-test/Test/Consensus/Cardano/Translation.hs:54]:
```suggestion
import           Ouroboros.Consensus.Ledger.Tables.Diffs (Diff)
import qualified Ouroboros.Consensus.Ledger.Tables.Diffs as Diff
```

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/MapKind.hs [hunk 4]
```diff
@@ -142,7 +142,7 @@ instance ZeroableMK DiffMK where
   emptyMK = DiffMK mempty
 
 instance CanMapKeysMK DiffMK where
-  mapKeysMK f (DiffMK (Diff.Internal.Diff m)) = DiffMK . Diff.Internal.Diff $
+  mapKeysMK f (DiffMK (Diff.Diff m)) = DiffMK . Diff.Diff $
```
Comment (jorisdral) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/MapKind.hs:145]:
```suggestion
  mapKeysMK f (DiffMK (Diff m)) = DiffMK . Diff $
```

File: ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore.hs [hunk 5]
```diff
@@ -332,18 +328,17 @@ instance (Ord k, QC.Arbitrary k, QC.Arbitrary v)
 deriving newtype instance (Ord k, QC.Arbitrary k, QC.Arbitrary v)
                        => QC.Arbitrary (Diff.Diff k v)
 
-instance QC.Arbitrary v => QC.Arbitrary (Diff.DeltaHistory v) where
+{- instance QC.Arbitrary v => QC.Arbitrary (Diff.DeltaHistory v) where
   arbitrary = Diff.DeltaHistory <$> ((:<||) <$> QC.arbitrary <*> QC.arbitrary)
   shrink (Diff.DeltaHistory h) =
     fmap Diff.DeltaHistory $ mapMaybe NESeq.nonEmptySeq $ QC.shrink (NESeq.toSeq h)
-
+ -}
```
Comment (jorisdral) [inline ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore.hs:335]: Remove

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Diffs.hs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+-- | Traversal with keys over the last delta in each delta history.
+traverseDeltaWithKey_ ::
+     Applicative t
+  => (k -> Delta v -> t a)
+  -> Diff k v
+  -> t ()
+traverseDeltaWithKey_ f (Diff m) = void $ Map.traverseWithKey f m
+
+-- | @'foldMap'@ over the last delta in each delta history.
```
Comment (jorisdral) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Diffs.hs:223]:
```suggestion
-- | @'foldMap'@ over the deltas.
```

**Review by jorisdral — APPROVED** [review]

**Pull request merged**